fenwick/src/main/java/org/opencadc/fenwick/InventoryHarvester.java

```diff
@@ -185,43 +194,108 @@ public InventoryHarvester(Map<String, Object> daoConfig, URI resourceID, Artifac
      */
     @Override
     public void run() {
+        boolean retry;
+        int retryCount = 1;
+        int sleepSeconds = INITIAL_SLEEP_TIMEOUT;
+
         while (true) {
             try {
                 final Subject subject = SSLUtil.createSubject(new File(CERTIFICATE_FILE_LOCATION));
-                Subject.doAs(subject, (PrivilegedExceptionAction<Void>) () -> {
-                    doit();
-                    return null;
+                int retries = retryCount;
+                int timeout = sleepSeconds;
+
+                retry = Subject.doAs(subject, (PrivilegedAction<Boolean>) () -> {
+                    boolean isRetry = true;
+                    try {
+                        doit();
+                        isRetry = false;
+                        // catch exceptions resulting in a retry
+                    } catch (ResourceNotFoundException ex) {
+                        // 404 service not found by TapClient, retry.
+                        logRetry(retries, timeout, "HTTP 404 resource not found", ex.getMessage());
+                    } catch (PreconditionFailedException ex) {
+                        // 412 precondition failed, retry.
+                        logRetry(retries, timeout, "HTTP 412 precondition failed", ex.getMessage());
+                    } catch (ByteLimitExceededException ex) {
+                        // 413 entity too large, retry.
+                        logRetry(retries, timeout, "HTTP 413 byte limit exceeded", ex.getMessage());
+                    } catch (ExpectationFailedException ex) {
+                        // 417 expectation failed, retry
+                        logRetry(retries, timeout, "HTTP 417 expectation failed", ex.getMessage());
+                    } catch (RemoteServiceException ex) {
+                        // 500 internal service error, retry.
+                        logRetry(retries, timeout, "HTTP 500 runtime error", ex.getMessage());
+                    } catch (TransientException ex) {
+                        //  503 transient, retry.
+                        logRetry(retries, timeout, "HTTP 503 transient", ex.getMessage());
+                    } catch (IOException ex) {
+                        // IO error talking to service, retry.
+                        logRetry(retries, timeout, "IO error", ex.getMessage());
+                    } catch (IllegalArgumentException ex) {
```

Comment:
Practical issue: the IllegalArgumentException below (part of the streaming fail hack) is ambiguous. The hack is catching this from the client side RowMapper that gets an incomplete row, but at end there is a comment that it is a service 400 error (because cadc-rest and cadc-util map IllegalArgumentException -> 400 -> IllegalArgumentException: so, you cannot tell if that is the service failing or the client failing. So, this exception should not be in the "keep trying" list at all. 

I think we need to retain detecting the truncated streaming but that probably needs to be done in the TapClient or RowMapper impls. I feel like the TapClient could be checking row data before handing off to RowMapper and it knows if end-of-stream was reached so should be able to detect end-of-stream and incorrect row data better than this. Need to discuss.

Reply:
Updated the TapClient to let it throw an IndexOutOfBoundsException when it tries to read a truncated stream.